src/inlineSuggestions/snippets/snippetProvider.ts

```diff
@@ -1,26 +1,58 @@
 import { Position, TextDocument } from "vscode";
 import {
+  AutocompleteResultState,
+  getAllSuggestions,
   getCurrentSuggestion,
   setSuggestionsState,
 } from "../inlineSuggestionState";
 import runCompletion from "../../runCompletion";
 import setInlineSuggestion from "../setInlineSuggestion";
 import { SnippetRequestTrigger } from "../../binary/requests/requests";
+import { sendEvent } from "../../binary/requests/sendEvent";
+
+enum SnippetEvents {
+  SnippetCanceledBeforeShown = "SnippetCanceledBeforeShown",
+  SnippetShown = "SnippetShown",
+}
 
 export default async function requestSnippet(
   document: TextDocument,
   position: Position,
   trigger: SnippetRequestTrigger = SnippetRequestTrigger.User
 ): Promise<void> {
-  const autocompleteResult = await runCompletion(
-    document,
-    position,
-    "snippet",
-    trigger
-  );
+  const autocompleteResult:
+    | AutocompleteResultState
+    | undefined
+    | null = await runCompletion(document, position, "snippet", trigger);
+
+  if (autocompleteResult) autocompleteResult.onStateChange = onStateChange;
```

Comment:
why the onStateChange is part of the autocompleteResult? it is not related to the autocompleteResult state

Reply:
its the `onStateChange` of this specific results batch, if for example the inline suggestions are setting the state with new results, i want the `onStateChange` to be removed

Reply:
if the state change is provided via a getter function, who would know when to call this function?
it should be called upon `alt+[ / ]`, creation of a new state and clearing of a state. i think that the best way to implement this is passing a callback that the inlineCompletionState module knows when to call, and the callback knows what to do.